File: packages/engine/src/cpus/6502/instructions/lsr.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('lsr acc', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.acc = 255
  mmu.data = [0x4a] // LSR
  const g = cpu.run()
  for (let i = 0; i < 2; i++) g.next()
  expect(cpu.acc).toBe(127)
  expect(cpu.carry).toBeTruthy()
})

test('lsr zpg', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x46, 0x02, 4] // LSR $02
  const g = cpu.run()
  for (let i = 0; i < 5; i++) g.next()
  expect(mmu.data[2]).toBe(2)
})

test('lsr abs', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  mmu.data = [0x4e, 0x02, 0xff] // LSR $ff02
  mmu.data[0xff02] = 4
  const g = cpu.run()
  for (let i = 0; i < 6; i++) g.next()
  expect(mmu.data[0xff02]).toBe(2)
})

File: packages/engine/src/cpus/6502/instructions/lsr.test.zig
const testing = @import("../testing/setup.zig");

test "LSR acc" {
    var cpu = testing.createCPU();
    cpu.acc = 255;
    cpu.write(0, 0x4a);
    for (0..2) |_| {
        cpu.process();
    }
    try testing.expectEqual(127, cpu.acc);
    try testing.expectEqual(true, cpu.c);
}

test "LSR zpg" {
    var cpu = testing.createCPU();
    cpu.write(0, 0x46);
    cpu.write(1, 2);
    cpu.write(2, 4);
    for (0..5) |_| {
        cpu.process();
    }
    try testing.expectEqual(2, cpu.read(2));
}

test "LSR abs" {
    var cpu = testing.createCPU();
    cpu.write(0, 0x4e);
    cpu.write(1, 2);
    cpu.write(2, 0xff);
    cpu.write(0xff02, 4);
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(2, cpu.read(0xff02));
}
